Development practices/Turbo frames.md

```diff
@@ -0,0 +1,195 @@
+
+Example:
+
+If we have a part of the page defined like this:
+
+```Html
+  ...
+  <turbo-frame id='id'>
+    <a href='/users/new'>New user</a>
+  </turbo-frame>
+  ...
+```
+
+Once the link to add a new user is clicked, the response might look like this:
+
+```Html
+  <body>
+    <h1> New user </h1>
+
+    <turbo-frame id='id'>
+      <form action='/users'>
+        <input name='user[name]' type='text' value='Name'>
+        <input type="submit">
+      </form>
+    </turbo-frame>
+  </body>
+```
+
+In this case, the frame from the response will be extracted and it will replace the frame with the same id from where the click originated.
+That means that the form for adding a new user will replace the link that was clicked and the rest of the page will remain intact.
+
+## Installation
+To be able to use the turbo frames in our application, the [turbo-rails](https://github.com/hotwired/turbo-rails) gem needs to be installed.
+
+The gem is automatically configured for applications made with Rails 7+ unless --skip-hotwire was passed to the generator.
+
+It can also be installed manually:
+1. Add the turbo-rails gem to your Gemfile: `gem 'turbo-rails'`
+2. Run `./bin/bundle install`
+3. Run `./bin/rails turbo:install`
+
+If you installed turbo to an existing application, it might be a good idea to disable it, and only enable it when needed.
+
+To disable turbo, add the following to one of the js files:
+```javascript
+import { Turbo } from "@hotwired/turbo-rails";
+Turbo.session.drive = false;
+```
+
+Then if you want to use turbo somewhere, just add `data-turbo="true"` to the relevant link or form.
+
+## Usage
+For this example we have an authors index page:
+```ruby
+  h2 = 'Authors'
+
+  - @authors.each do |author|
+    .row
+      .col
+        = author.first_name
+      .col
+        = author.last_name
+      .col
+        = link_to 'Edit', edit_author_path(author.id)
+```
+
+There is a list of authors and next to each author there is an edit button.
+Now let's say that we want to be able to edit the authors inline, without redirecting to a new page.
+
+What would usually require some javascript code, can be easily done using turbo frames with just a few changes on the existing code:
+```ruby
+  h2 = 'Authors'
+
+  - @authors.each do |author|
+    = turbo_frame_tag "edit_author_#{author.id}"
+      .row
+        .col
+          = author.first_name
+        .col
+          = author.last_name
+        .col
+          = link_to 'Edit', edit_author_path(author.id)
+```
+
+The only required change was adding this line for each author:
+```ruby
+  = turbo_frame_tag "edit_author_#{author.id}"
+```
+
+The `turbo_frame_tag` helper method will wrap the content in a `<turbo-frame>` tag.
+
+For example, for the author with the `id=1` it will wrap the block of code in `<turbo-frame id="edit_author_1">` tag.
+
+The view for the #edit action may look like this:
+```ruby
+  = turbo_frame_tag "edit_author_#{@author.id}"
```

Comment:
what if we want to use `render 'form'`? In that case, it wouldn't work because `new` and `edit` would have a different frame id.

Please try with the following and add it as a section in this chapter if it works (eg: Working with partials for forms):

```ruby
= turbo_frame_tag [action_name, dom_id(@author)]
```

Reply:
This does not work because the `dom_id(@author)` for the #new action is `new_author` so
`= turbo_frame_tag [action_name, dom_id(@author)]` would generate the turbo frame with the `new_new_author` id.

Also if there some validation errors while editing, the action_name would be `update` instead of `edit`.